File: 206.reverse-linked-list.cpp
/*
 * @lc app=leetcode id=206 lang=cpp
 *
 * [206] Reverse Linked List
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        auto cur=head;
        ListNode* prev_node=nullptr;
        while(cur!=nullptr){
            auto next_cur=cur->next;
            cur->next=prev_node;
            prev_node=cur;
            cur=next_cur;
        }
        return prev_node;
    }
};
// @lc code=end

File: java/206.reverse-linked-list.java
/*
 * @lc app=leetcode id=206 lang=java
 *
 * [206] Reverse Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode() {} ListNode(int val) { this.val = val; } ListNode(int val,
 * ListNode next) { this.val = val; this.next = next; } }
 */
class Solution {
    // public class ListNode {
    //     int val;
    //     ListNode next;

    //     ListNode() {
    //     }

    //     ListNode(int val) {
    //         this.val = val;
    //     }

    //     ListNode(int val, ListNode next) {
    //         this.val = val;
    //         this.next = next;
    //     }
    // }
    public ListNode reverseList(ListNode head) {
        ListNode prev_node = null;
        var cur = head;
        while (cur != null) {
            var next_cur = cur.next;
            cur.next = prev_node;
            prev_node = cur;
            cur = next_cur;
        }
        return prev_node;
    }
}
// @lc code=end
